### PR TITLE
[SPARK-31407][SQL][TEST] TestHiveQueryExecution should respect database when creating table

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

In `TestHiveQueryExecution`, if we detect a database in the referenced table, we should create the table under that database.

### Why are the changes needed?

This fix the test `Fix hive/SQLQuerySuite.derived from Hive query file: drop_database_removes_partition_dirs.q` which currently only pass when we run it with the whole test suit but fail when run it separately.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Run the test separately and together with the whole test suite.

### CODE REVIEW

**Comment by Ngone51** [comment]
cc @cloud-fan

**Review by HyukjinKwon — COMMENTED** [review]
Why don't we add `sql("SELECT count(*) FROM src").queryExecution.analyzed` at the very beginning of the test so it can initialize the table via `TestHiveQueryExecution.analyzed` in default database?

**Comment by Ngone51** [comment]
> Why don't we add sql("SELECT count(*) FROM src").queryExecution.analyzed at the very beginning of the test so it can initialize the table via TestHiveQueryExecution.analyzed in default database?

So what about others tables?  Initialize `src` table only just looks a bit weird to me. If we want to create tables before any tests we can simply create in `TestHiveQueryExecution` instead of current way?

Another alternative way is to do something like below in `TestHiveQueryExecution` if it pass all tests:

```
try {
 // set currentDB to default
 loadTestTable(some_table)
} finally {
 // restore original currentDB
}
```

WDYT?

**Comment by HyukjinKwon** [comment]
Are there multiple tests failed for this reason? If not, I think we can just add one liner for the workaround for now. The test seems like targeting to test the access of different databases with different default apparently. If there's an easy workaround to keep the test as is, let's do that.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 1]
```diff
@@ -588,20 +588,29 @@ private[hive] class TestHiveQueryExecution(
 
   override lazy val analyzed: LogicalPlan = {
     val describedTables = logical match {
-      case CacheTableCommand(tbl, _, _, _) => tbl.table :: Nil
+      case CacheTableCommand(tbl, _, _, _) => (tbl.table, tbl.database) :: Nil
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala:591]: nit, what about we just use `TableIdentifier` as is?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 2]
```diff
@@ -588,20 +588,29 @@ private[hive] class TestHiveQueryExecution(
...
       case _ => Nil
     }
 
     // Make sure any test tables referenced are loaded.
     val referencedTables =
       describedTables ++
-        logical.collect { case UnresolvedRelation(ident) => ident.last }
+        logical.collect { case UnresolvedRelation(ident) => (ident.last, ident.init.lastOption) }
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala:598]: and .. we `unapply` with `AsTableIdentifier` or call `IdentifierHelper.asTableIdentifier`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 3]
```diff
@@ -588,20 +588,29 @@ private[hive] class TestHiveQueryExecution(
...
     val resolver = sparkSession.sessionState.conf.resolver
-    val referencedTestTables = sparkSession.testTables.keys.filter { testTable =>
-      referencedTables.exists(resolver(_, testTable))
+    val referencedTestTables = referencedTables.flatMap { case (table, dbOpt) =>
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala:600]: and we iterate here with `TableIdentifier` instance.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 4]
```diff
@@ -588,20 +588,29 @@ private[hive] class TestHiveQueryExecution(
...
+      val testTableOpt = sparkSession.testTables.keys.find(resolver(_, table))
+      testTableOpt.map(testTable => Seq((testTable, dbOpt))).getOrElse(Seq.empty)
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala:602]: nit: I think we can just do `testTableOpt.map(testTable => (testTable, tbl.database))`

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise

**Comment by HyukjinKwon** [comment]
Don't forget to update PR description and title :).

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala:591]: +1

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala (hunk 2 above)
Comment (Ngone51) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala]: Seems it's not convenient and proper to use either of them. `AsTableIdentifier` requires `TestHiveQueryExecution` to extend `LookupCatalog` and implement `catalogManager`. And `IdentifierHelper` actually requires `IdentifierImpl` instead.

So I used `TableIdentifier` directly.

**Comment by Ngone51** [comment]
Thanks a lot @HyukjinKwon , I've updated description and title.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 5]
```diff
@@ -588,20 +589,31 @@ private[hive] class TestHiveQueryExecution(
...
+      case CacheTableCommand(tbl, _, _, _) => tbl :: Nil
       case _ => Nil
     }
 
     // Make sure any test tables referenced are loaded.
     val referencedTables =
       describedTables ++
-        logical.collect { case UnresolvedRelation(ident) => ident.last }
+        logical.collect { case UnresolvedRelation(ident) =>
+          TableIdentifier(ident.last, ident.init.lastOption)
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala:600]: we should assert that `ident.length <= 2`

**Review by cloud-fan — APPROVED** [review]

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala (hunk 5 above)
Comment (HyukjinKwon) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala:600]: Or you can just use:

```scala
import org.apache.spark.sql.connector.catalog.CatalogV2Implicits._
...
logical.collect { case UnresolvedRelation(ident) => ident.asTableIdentifier }
```

I guess.

Comment (Ngone51) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/test/TestHive.scala:600]: I see. this one should work.

**Comment by Ngone51** [comment]
retest this please.

**Comment by Ngone51** [comment]
retest this please.

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.0.

**Comment by Ngone51** [comment]
thanks all!

**Pull request closed without merging**